### DISCUSSION TITLE
Enabling GPUS in CIME

### DISCUSSION THREAD

**Discussion by jedwards4b** [ESMCI/cime#4333] [Show and tell]
I have created a cesm branch in my fork called gpu_improvements:

git clone https://github.com/jedwards4b/CESM.git -b gpu_improvements
cd cesm./manage_externals/checkout_externals

There are two new options to create_newcase --gpu-type and --gpu-offload
I have reworked the tests that you defined in cam to use smaller resolutions and also the se dycore and different numbers of gpus per node.The compiler is nvhpc not nvhpc-gpu - compiler definitions with -gpu will be removed by the time this PR is accepted.   I am also going to share this 
with the e3sm team for feedback prior to merging so there may be further changes.  

One question I have is what to do if there is more than one offload method in the model, one section using OpenACC and another using OpenMP?I'm not sure how we would handle that.

**Comment by jedwards4b** [comment]
See changes in cime here: https://github.com/ESMCI/cime/pull/4334

**Comment by jedwards4b** [comment]
Changes in other components are here:
https://github.com/ESMCI/ccs_config_cesm/pull/66
https://github.com/ESCOMP/CMEPS/pull/318
https://github.com/ESCOMP/CAM/pull/682

**Comment by rljacob** [comment]
Can you provide some create_newcase useage examples?   Or a machine file example?

**Reply by jedwards4b** [reply]
Here is a create_test example:
``` 
./create_test SMS_Ln9_G4-v100-OpenACC.ne16_ne16_mg17.F2000dev.casper_nvhpc.cam-outfrq9s_mg3_default
```
Which results in a create_newcase:
```
create_newcase.py --case /glade/scratch/jedwards/SMS_Ln9_G4-v100-OpenACC.ne16_ne16_mg17.F2000dev.casper_nvhpc.cam-outfrq9s
_mg3_default.20221103_080018_myf2kj --res ne16_ne16_mg17 --compset F2000dev --test --machine casper --compiler nvhpc --project P93300606 --workflow default --srcroot /glade/work/jedwards/sandboxes/cesm2_x_a
lpha --ngpus-per-node 4 --gpu-type v100 --gpu-offload OpenACC --user-mods-dir /glade/work/jedwards/sandboxes/cesm2_x_alpha/components/cam//cime_config/testdefs/testmods_dirs/cam/outfrq9s_mg3_default
```

**Reply by bartgol** [reply]
When running `crete_newcase` (rather than `create_test`), does one have to manually specify the gpu type? Can't it be retrieved from machine specs?

**Reply by jedwards4b** [reply]
we have machines with multiple nodes with different gpu types and multiple compilers not all of which support gpus.  So no I don't think we can do that.

**Reply by bartgol** [reply]
Can it at least have a "default" value? Some machines have a single gpu type, so it can be handy to have it hard coded in config_machines.xml, and save the user a few key strokes (when possible).

**Reply by jedwards4b** [reply]
Another case is that a user wants to use the given compiler and machine but does not want to use gpus.   I'll keep this suggestion in mind, but I don't know of a way to do it at the moment.

**Reply by bartgol** [reply]
In that case, I would expect the user to not pass `--gpu-offload`, which would make CIME ignore any other GPU setting.

Anyhow, this is just a super minor thing.

**Comment by sjsprecious** [comment]
Thanks @jedwards4b for working on this! Regarding your question, I am not sure whether it is worth the effort to consider the case that both OpenACC and OpenMP offload are used at the same time. If you do want to account for this type of situation, we can apply the compilation flags `-acc -gpu=cc70 -mp=gpu` for the NVHPC compiler, which would compile both OpenACC and OpenMP offload codes at the same time. However, in my opinion, by doing so, we probably do not need the `gpu_offload` variable anymore.

**Reply by jedwards4b** [reply]
So we could define an offload method 'combined' which would support both.

**Reply by sjsprecious** [reply]
That sounds like a good option!

**Comment by jedwards4b** [comment]
From @sarats : As this mandates GPU_OFFLOAD, we probably have to add a third option to handle the case for C++ codebases that use a library approach as it won't be OpenMP or OpenACC.

**Reply by sarats** [reply]
The `combined` mode would also help C++ components built with Fortran+Directives componenents in a coupled configuration. But one has to distinguish between OpenACC and OpenMP within the "combined" conditionals if needed to define flags.  There will always be configuration complexity (with or without GPU_OFFLOAD var) to handle special cases, so it may not lead to concision in those cases.

**Comment by sarats** [comment]
Can you illustrate an example of where specialization for GPU_TYPE resides? In CIME machine section?

**Reply by jedwards4b** [reply]
I'm not sure if I understand the question but for cesm we have cmake flags in: https://github.com/ESMCI/ccs_config_cesm/pull/66/files#diff-e9435a30e3faf12b3429137c9ac98de1a71394ad1c3dc17f284af32d9f8cf4b8R52

**Comment by oksanaguba** [comment]
From e3sm perspective, is it needed in e3sm? right now we have cmake macros files for "compiler"/machine pairs (where "compiler" can be any config we want to support) with flexible cmake vars, including conditional ones for each component. So, would this PR add functionality?

**Reply by jedwards4b** [reply]
Maybe the question is - is it going to adversely affect current e3sm functionality?

**Reply by bartgol** [reply]
I think putting the threading model in either "machine" or "compiler" category is suboptimal, since it can cause duplication of machines, and/or confusion. That said, we are already doing it, so it is a solution.

**Reply by jedwards4b** [reply]
I don't understand this comment - this PR elements the need to define a special compiler for enabling gpu's and reduces duplication.

**Reply by bartgol** [reply]
I think what oksana is suggesting is to make a new machine (or compiler) for each offload choice, and in those machines/compilers we can set specific compiler flags. We have done that in E3SM for some cpu/gpu machines.

**Reply by jedwards4b** [reply]
Again that is what I am trying to undo - we don't need to completely duplicate compiler definitions based on whether or not we want to use gpus.

**Reply by bartgol** [reply]
I agree. Although, if we came up with a way to "inherit" XML nodes, we could have
```
<machine MACH="summit-base">
  all common settings
</machine>

<machine MACH="summit-gpu" inherit="summit-base">
  gpu specific settings
</machine>
```
This would not be completely flexible, for it would still have duplication in case 2+ gpu archs need to be used at the same time (I wasn't aware this was ever the case, btw, I always thought one GPU arch at a time).

**Reply by bartgol** [reply]
So, tbc, I agree with oksana that we can already get this behavior by multiplying mach files. But I agree that we should split the threading model concept from the machine/compiler specs.

**Reply by rljacob** [reply]
What we decided in E3SM:  for machines like perlmutter which provide one front end to two different node types (cpu-only nodes and cpu-gpu nodes), we have two machine entries.   For machines like Summit where there is only one type of heterogeneous node, we have one machine entry but two compiler entries:  one for cpu-only compile and one for gpu compile.   As long as we can keep doing that, this won't impact E3SM.

**Reply by rljacob** [reply]
I see how the one-machine-entry-per-node-type can get very duplicative when you have N>4 node types hanging off a single front end.

**Comment by bartgol** [comment]
I wonder if GPU and CPU threading can be unified in a single option. Something like `--threading=XYZ`, with possible XYZ values `none`, `openmp`, `openmp-target`, `openacc`, `cuda`, `hip`,... Maybe even a list of those, to enable, say, OpenMP on host, and both openacc and cuda on device (if some components use the former while others use the latter).

**Reply by jedwards4b** [reply]
This is an interesting idea.   Currently we enable openmp with the NTHRDS_{_COMP_} variable which can be set individually for each component.  We could build on this with THREADTYPE_{_COMP_} instead of the GPU_OFFLOAD variable I have introduced.  In this case the NTHRDS variable would only apply to host threading and we still need NGPUS_PER_NODE for offloading.

**Reply by bartgol** [reply]
This would also allow to specify a different threading model for each component (e.g., run ATM with cuda, but lnd with openacc, and the rest with classic openmp).

**Reply by jedwards4b** [reply]
@sjsprecious what do you think?

**Reply by sjsprecious** [reply]
Hmm, interesting. Let's say PUMAS is running on GPU with OpenACC and RRTMGP is running on GPU with OpenMP offload. Both components are in ATM. How could this idea work then? lf we need to do something like `--threading=openacc,openmp-target`, basically it is equivalent to set `GPU_OFFLOAD=combined` in my opinion and I do not see the benefit of using this opinion here. Plus linking the term `threading` to `openacc`, `openmp-target`, `cuda` and `hip` looks weird to me.

**Reply by sjsprecious** [reply]
And I think EAM has more parameterizations (probably all?) running on GPU so they are able to test at the component level. But for CAM, only a few parts are ported to GPU. Thus what I think works best for CAM is to let the user specify what GPU types and offload methods they want. Then CIME is able to apply the correct GPU flags to the selected files, build a case and submit the job to the right queue. And I believe that is what you are doing right now in your PRs.

**Reply by bartgol** [reply]
Wh does it sound weird? GPUs _do_ use threads.

**Reply by bartgol** [reply]
> And I think EAM has more parameterizations (probably all?) running on GPU so they are able to test at the component level. But for CAM, only a few parts are ported to GPU. Thus what I think works best for CAM is to let the user specify what GPU types and offload methods they want. Then CIME is able to apply the correct GPU flags to the selected files, build a case and submit the job to the right queue. And I believe that is what you are doing right now in your PRs.

This points to a different issue: should CIME be in charge at all of GPU settings? Or should the component do its own internal choices?

**Reply by jedwards4b** [reply]
CIME was develop in part because allowing individual components too much autonomy in setting compiler flags lead to contradictions and conflicts that made it difficult to put a multi-component executable together.  CIME is in charge but leaves
room for components to customize as needed.

**Reply by bartgol** [reply]
> > And I think EAM has more parameterizations (probably all?) running on GPU so they are able to test at the component level. But for CAM, only a few parts are ported to GPU. Thus what I think works best for CAM is to let the user specify what GPU types and offload methods they want. Then CIME is able to apply the correct GPU flags to the selected files, build a case and submit the job to the right queue. And I believe that is what you are doing right now in your PRs.
> 
> This points to a different issue: should CIME be in charge at all of GPU settings? Or should the component do its own internal choices?

I guess I am not thinking too much in CIME terms, since I'm too used to target-based cmake. But here, components are still built centrally by CIME.

How does CIME distringuish between different parts of CAM, when it's time to apply flags? Does it apply, say, both openmp-target and cuda flags to all CAM files? Or is there a mechanism to know which ones need which?

**Reply by jedwards4b** [reply]
We use the Depends.* feature to specify files.   Here is an example:
cat Depends.nvhpc
```
PUMAS_OBJS=\                                                                                                                                                                                                  
micro_mg1_0.o \                                                                                                                                                                                               
micro_pumas_v1.o \                                                                                                                                                                                            
micro_pumas_data.o \                                                                                                                                                                                          
micro_pumas_utils.o \                                                                                                                                                                                         
micro_pumas_cam.o \                                                                                                                                                                                           
wv_sat_methods.o \                                                                                                                                                                                            
wv_saturation.o \                                                                                                                                                                                             
macrop_driver.o \                                                                                                                                                                                             
shr_spfn_mod.o                                                                                                                                                                                                
                                                                                                                                                                                                              
$(PUMAS_OBJS): %.o: %.F90                                                                                                                                                                                     
        $(FC) -c $(INCLDIR) $(INCS) $(FFLAGS) $(FREEFLAGS) $(GPUFLAGS) $<                                                                                                                                     
                                                                                                         
```

**Reply by sjsprecious** [reply]
> Wh does it sound weird? GPUs _do_ use threads.

Sorry that I may not use an appropriate word here. The people I worked with always treated `threading` as OpenMP on CPU. And we used the term `OpenMP offload` for using OpenMP on GPU to distinguish. Talking about OpenACC, I do not think there is a clause `thread` in OpenACC until it is translated to CUDA. Thus if we do want a single option here, I would suggest something more general like `--gpu_offload` rather than `--threading` here. This is my point.

**Reply by jedwards4b** [reply]
I actually think that threading is more general than gpu_offload.  But I suggest we implement with gpu_offload for now and perhaps consider this change for later.

**Reply by bartgol** [reply]
Ok, thanks for the clarification!

**Reply by sjsprecious** [reply]
And one more question: does E3SM use OpenACC or OpenMP offload for GPU? I thought E3SM used KoKKos for GPU offload so this change may not be used in E3SM at all? Or maybe we should add `kokkos` as a valid `gpu_offload` value?

**Reply by jedwards4b** [reply]
@sjsprecious I think that it would be cuda not kokkos, but e3sm uses a different mechanism so I don't think we need to worry about that yet.

**Reply by bartgol** [reply]
Well, E3SM can currently use several ways. Kokkos is one. YAKL is a similar (smaller) library. But other components still use openacc, I think. Either way, I think E3SM is moving toward a more cmake-target-centric approach (at least I hope it will). But I would still be interested in grabbing threading model specs from the CIME case configuration (rather than rely on simply machine files). I think, at least for the ATM, what we would need to make our build decision is knowing what kind of device we're using. After that, we can configure things internally via CMake.

**Reply by sjsprecious** [reply]
Thanks @jedwards4b and @bartgol for your clarifications. That helps a lot! Sounds like the ATM component in E3SM can fully run on GPU now?

**Reply by bartgol** [reply]
Not in current release of E3SM. But yes, we have a version of ATM for ultra high res that is fully running on GPU end to end (except I/O, obviously).
